### PR TITLE
MDL-72397 qbank_managecategories - UI enhancement

### PR DESCRIPTION
**Pull request by marcghaly**
UI enhancement for qbank_managecategories, files that were altered or created are the following:

- question/bank/managecategories/amd/build/addcategory_dialogue.min.js
- question/bank/managecategories/amd/build/addcategory_dialogue.min.js.map
- question/bank/managecategories/amd/build/checkbox_param.min.js
- question/bank/managecategories/amd/build/checkbox_param.min.js.map
- question/bank/managecategories/amd/build/order_categories.min.js
- question/bank/managecategories/amd/build/order_categories.min.js.map
- question/bank/managecategories/amd/src/addcategory_dialogue.js
- question/bank/managecategories/amd/src/checkbox_param.js
- question/bank/managecategories/amd/src/order_categories.js
- question/bank/managecategories/category.php
- question/bank/managecategories/classes/external/set_category_order.php
- question/bank/managecategories/classes/external/submit_add_category_form.php
- question/bank/managecategories/classes/form/question_category_checkbox_form.php
- question/bank/managecategories/classes/form/question_category_edit_form.php
- question/bank/managecategories/classes/question_category_list.php
- question/bank/managecategories/classes/question_category_list_item.php
- question/bank/managecategories/classes/question_category_object.php
- question/bank/managecategories/db/services.php
- question/bank/managecategories/lang/en/qbank_managecategories.php
- question/bank/managecategories/lib.php
- question/bank/managecategories/pix/grip-vertical-solid.svg
- question/bank/managecategories/templates/basecategory.mustache
- question/bank/managecategories/templates/categorylist.mustache
- question/bank/managecategories/templates/categoryobject.mustache
- question/bank/managecategories/templates/listitem.mustache
- question/bank/managecategories/tests/behat/category_reorder.feature
- question/bank/managecategories/tests/managecategories_enhancement_test.php
- question/bank/managecategories/tests/question_category_object_test.php
- question/tests/events_test.php

### CODE REVIEW

File: question/bank/managecategories/amd/src/addcategory_dialogue.js [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * Javascript module handling creation of Modal form.
+ *
+ * @package    qbank_managecategories
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ */
+
+import $ from 'jquery';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/amd/src/addcategory_dialogue.js]: fixed

File: question/bank/managecategories/amd/src/addcategory_dialogue.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    modal.getRoot().find('.error')
+  );
+
+  // If we found invalid fields, focus on the first one and do not submit via ajax.
+  if (invalid.length) {
+      invalid.first().focus();
+      return;
+  }
+
+  let formData =  modal.getRoot().find('form').serialize();
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/amd/src/addcategory_dialogue.js]: fixed

File: question/bank/managecategories/amd/src/order_categories.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+const setupSortableLists = () => {
+    new SortableList(
+        '.category_list',
+        {
+            moveHandlerSelector: '.list_item [data-drag-type=move]',
+        }
+    );
+    $('.list_item').on(SortableList.EVENTS.DROP, (evt) => {
+        evt.stopPropagation();
+        let categoryListElements = $('.list_item').parent();
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/amd/src/order_categories.js]: replaced by `const` where possible

File: question/bank/managecategories/category.php [hunk 4]
```diff
@@ -24,78 +24,50 @@
...
 
 require_login();
 core_question\local\bank\helper::require_plugin_enabled(helper::PLUGINNAME);
 
 list($thispageurl, $contexts, $cmid, $cm, $module, $pagevars) =
         question_edit_setup('categories', '/question/bank/managecategories/category.php');
 
+$thiscontext = context_module::instance($cmid)->id;
+
+// Javascript calls.
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/category.php:43]: Fixed and replaced with imports

File: question/bank/managecategories/classes/external/submit_add_category_form.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            ]);
+    }
+
+    /**
+     * Submit the add category form.
+     *
+     * @param string $jsonformdata The data from the form, encoded as a json array.
+     * @return int $categoryid category id.
+     */
+    public static function execute($jsonformdata) {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/submit_add_category_form.php:60]: fixed

File: question/bank/managecategories/classes/external/set_category_order.php [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+        $oldctxid = (int)explode(' ', $categories[2])[0];
+        $newctxid = (int)explode(' ', $categories[1])[0];
+
+        // Question_categories table modifications.
+        if (!is_null($categories[1])) {
+            // Retrieves top category parent where neighbor category is located.
+            $destparentcat = $DB->get_record_select('question_categories', 'contextid = :newcontextid AND parent = :parentcat',
+                    ['newcontextid' => $newctxid, 'parentcat' => 0]);
+            // Sets new parent.
+            $DB->set_field('question_categories', 'parent', $destparentcat->id, ['id' => $catid]);
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/set_category_order.php]: Made my best to reduce DB queries to minimum.

**Comment by marcghaly** [comment]
Hi @mattporritt, conversations have been resolved,
Cheers !

**Comment by safatshahin** [comment]
Hi @marcghaly 
I havent done the full review but found some issues:
1. An error from context while accessing the page from course>question bank>questions, reason its just considering/forcing the cmid, which should also consider course id, as the access can happen from any place.
2. An error for parameter from question_category_list_item as the required param is cmid, that not right, it can come from any place, course/module.
Please resolved them so that I can do a full review.
Cheers

**Comment by marcghaly** [comment]
Hi @safatshahin, cmid/courseid error has been fixed in order to be accessible from anywhere, the `required_param` for `cmid` in `question_category_list_item` has been replaced and the unit test has been fixed.

Thank you

**Review by safatshahin — COMMENTED** [review]
Hi @marcghaly 
Good work on this, this page used to be a little here and there but now it looks way too clean.
And also this change takes it to a stage where anyone can make it prettier at any point in time by making some small changes in mustache and some CSS. 
I have made some comments, nothing horribly wrong, some bits and pieces. As soon as they are addressed and its rebased against master, it will be good to go for initial peer review.
There are a lot of xpath elements in the behat test, it's better to use a generator, but you're allowed to do that, and drag and drop are always a bit weird. So all good for now.
Cheers!

File: question/bank/managecategories/templates/basecategory.mustache [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+<div class="row">
+    <div class="pl-4">
+        {{{helpstringhead}}}
+    </div>
+</div>
+<br>
+<div class="row">
+    <div class="pl-4">
+        <button type="submit" data-action="addcategory" class="btn btn-primary">
```
Comment (safatshahin) [inline question/bank/managecategories/templates/basecategory.mustache:38]: need capability check around this button, for both webservice and mustache

File: question/bank/managecategories/amd/src/checkbox_param.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+ * @copyright  2021 Catalyst IT Australia Pty Ltd
+ * @author     Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ */
+class CheckboxParam {
+    /**
+     * Sets events listenner for checkbox ticking change
+     *
+     * @returns
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/amd/src/checkbox_param.js:29]: fixed

File: question/bank/managecategories/amd/src/checkbox_param.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    setEventListenner = () => {
+        let checkbox = document.getElementsByName('qbshowdescr')[0];
+        let checkboxform = document.getElementById('qbshowdescr-form');
+        checkbox.addEventListener('click', (e) => {
+            e.preventDefault();
+            checkboxform.submit();
+        });
+    };
+}
+export default CheckboxParam;
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/amd/src/checkbox_param.js:40]: fixed

File: question/bank/managecategories/classes/external/set_category_order.php [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    public static function execute(string $categories) {
+        global $DB;
+        $params = self::validate_parameters(self::execute_parameters(),
+            ['categories' => $categories]);
+        $context = context_system::instance();
+        self::validate_context($context);
+        require_capability('moodle/category:manage', $context);
+        // New order insertion.
+        $categories = json_decode($categories, true);
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/set_category_order.php:65]: fixed

File: question/bank/managecategories/classes/external/set_category_order.php [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+require_once($CFG->libdir . "/externallib.php");
+
+use context_system;
+use external_api;
+use external_function_parameters;
+use external_value;
+use qbank_managecategories\helper;
+
+class set_category_order extends external_api {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/set_category_order.php:40]: fixed

File: question/bank/managecategories/classes/external/submit_add_category_form.php [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+use context_system;
+use external_api;
+use external_function_parameters;
+use external_value;
+use context;
+use stdClass;
+use moodle_exception;
+
+class submit_add_category_form extends external_api {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/external/submit_add_category_form.php:42]: fixed

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 13]
```diff
@@ -16,7 +16,14 @@
 
 namespace qbank_managecategories;
 
+defined('MOODLE_INTERNAL') || die();
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/question_category_list_item.php:19]: fixed

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 14]
```diff
@@ -90,15 +97,76 @@ public function item_html($extraargs = []) : string {
...
+                    false
+                ));
+            }
         }
 
+        // Sets up export to XML link.
+        if (class_exists('\\qbank_exportquestions\\form\\export_form')) {
+            $exporturl = new moodle_url('/question/bank/exportquestions/export.php',
+                ['cat' => $category->id . ',' . $category->contextid]);
+        } else {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/question_category_list_item.php:147]: fixed

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 15]
```diff
@@ -90,15 +97,76 @@ public function item_html($extraargs = []) : string {
...
+            $exporturl = new moodle_url('/question/export.php',
+            ['cat' => $category->id . ',' . $category->contextid]);
+        }
+        if ($courseid !== 0) {
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/question_category_list_item.php:151]: fixed

File: question/bank/managecategories/classes/question_category_list_item.php [hunk 16]
```diff
@@ -90,15 +97,76 @@ public function item_html($extraargs = []) : string {
...
+            $exporturl->param('courseid', $courseid);
+        } else {
+            $exporturl->param('cmid', $cmid);
+        }
+        $menu->add(new action_menu_link(
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/classes/question_category_list_item.php:156]: fixed

File: question/bank/managecategories/db/services.php [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+ * @author     2021, Ghaly Marc-Alexandre <marc-alexandreghaly@catalyst-ca.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+$functions = [
+    'qbank_managecategories_set_category_order' => [
+        'classname'    => 'qbank_managecategories\external\set_category_order',
+        'classpath'    => '',
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/db/services.php:32]: fixed

File: question/bank/managecategories/db/services.php [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+        'description'  => 'Returns question category order',
+        'type'         => 'write',
+        'capabilities' => 'moodle/question:managecategory',
+        'ajax'         => true,
+    ],
+    'qbank_managecategories_submit_add_category_form' => [
+        'classname'    => 'qbank_managecategories\external\submit_add_category_form',
+        'classpath'    => '',
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/db/services.php:40]: fixed

File: question/bank/managecategories/lang/en/qbank_managecategories.php [hunk 19]
```diff
@@ -23,5 +23,12 @@
  * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
-$string['pluginname'] = 'Manage categories';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/lang/en/qbank_managecategories.php:26]: fixed

File: question/bank/managecategories/lang/en/qbank_managecategories.php [hunk 20]
```diff
@@ -23,5 +23,12 @@
...
-$string['privacy:metadata'] = 'The Manage Categories plugin does not store any user data.';
+$string['addcategory'] = 'Add category';
+$string['alertcheckboxes'] = '<h4> Please check at least one category to delete</h4>';
+$string['deleteallquestions'] = 'Delete all questions';
+$string['deletecategoryformname'] = 'Delete category';
+$string['deletecategorylabel'] = 'Delete';
+$string['movequestions'] = 'Move questions to category';
+$string['pluginname'] = 'Manage question bank categories feature';
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/lang/en/qbank_managecategories.php]: fixed

File: question/bank/managecategories/pix/grip-vertical-solid.svg [hunk 21]
```diff
@@ -0,0 +1 @@
+<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="grip-vertical" class="svg-inline--fa fa-grip-vertical fa-w-10" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><path fill="currentColor" d="M96 32H32C14.33 32 0 46.33 0 64v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160H32c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zM288 32h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32V64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32zm0 160h-64c-17.67 0-32 14.33-32 32v64c0 17.67 14.33 32 32 32h64c17.67 0 32-14.33 32-32v-64c0-17.67-14.33-32-32-32z"></path></svg>
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/pix/grip-vertical-solid.svg:1]: fixed

File: question/bank/managecategories/templates/categorylist.mustache [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+}}
+{{!
+    @template qbank_managecategories/categorylist
+
+    This template renders the list item for each category.
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/templates/categorylist.mustache:20]: fixed

File: question/bank/managecategories/templates/categoryobject.mustache [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+    Example context (json):
+    {
+        "ctxlvl": "70",
+        "heading": "Question categories for 'Category: Miscellaneous",
+        "list": "<ul class="category_list">...</ul>"
+    }
+}}
+
+<div class="box py-3 boxwidthwide boxaligncenter generalbox questioncategories contextlevel{{ctxlvl}}">
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/templates/categoryobject.mustache:30]: fixed

File: question/bank/managecategories/templates/basecategory.mustache [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+    </div>
+</div>
+<br>
+<div id="categoriesrendered">
+    {{#categoriesrendered}}
+        {{{category}}}
+    {{/categoriesrendered}}
+</div>
+<br>
+<br>
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/templates/basecategory.mustache:54]: fixed

File: question/bank/managecategories/tests/behat/category_reorder.feature [hunk 25]
```diff
@@ -0,0 +1,78 @@
+@qbank @qbank_managecategories @javascript
+Feature: A Teacher can reorder question categories
+  In order to change question category order
+  As a Teacher
+  I need to reorder them
+
+  Background:
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/tests/behat/category_reorder.feature:7]: I fixed a existing failing behats and one contained that test, should it be done in this file ?

File: question/bank/managecategories/tests/managecategories_enhancement_test.php [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+        $newcategories = $DB->get_records('question_categories');
+        $neworder = [];
+        foreach ($newcategories as $category) {
+            $neworder[] = [$category->contextid . ' ' . $category->id];
+        }
+        $ordertoset = json_decode($ordertoset);
+
+        $this->assertNotEquals($ordertoset[0], $neworder);
+    }
+}
```
[the remark this answers is not in the document]
Comment (marcghaly) [reply question/bank/managecategories/tests/managecategories_enhancement_test.php]: fixed

File: question/bank/managecategories/templates/basecategory.mustache (hunk 7 above)
Comment (marcghaly) [reply question/bank/managecategories/templates/basecategory.mustache]: fixed

**Comment by marcghaly** [comment]
Hi @safatshahin, all conversations resolution have been made. 
Thank you

**Comment by safatshahin** [comment]
Hi @marcghaly closing this PR and creating a new one because of commit hash issue with master

**Pull request closed without merging**